File: src/ts/supabase_sync.ts
import { Subject, filter, type Observable } from "rxjs";

import {
  commitAcceptedDrop,
  telemetryFor,
  type DndDropResult,
  type DndEnvelope,
  type DndTelemetryEvent,
  type DropCommitPorts,
  type OptoSyncPort,
  type OresFormsPort,
  type OresOtelPort,
} from "./index.js";
import { OresDndReactiveBus } from "./reactive.js";

export type DndSyncChannel = "opto-sync" | "ores-otel";

export interface DndSupabaseSyncReceipt {
  readonly dragId: string;
  readonly channel: DndSyncChannel;
  readonly backend: "supabase";
  readonly ok: boolean;
  readonly targetId?: string;
  readonly errorCode?: "sync-failed";
}

/** Host adapter: Opto-Sync remains the local-first/state synchronization owner. */
export interface OptoSyncSupabasePort extends OptoSyncPort {
  syncAcceptedDropToSupabase(
    envelope: DndEnvelope,
    result: DndDropResult,
  ): void | Promise<void>;
}

/** Host adapter: only sanitized ORES-OTel metadata crosses this boundary. */
export interface OresOtelSupabasePort extends OresOtelPort {
  syncDndEventToSupabase(event: DndTelemetryEvent): void | Promise<void>;
}

export interface SupabaseDropCommitPorts {
  readonly forms?: OresFormsPort;
  readonly optoSync?: OptoSyncSupabasePort;
  readonly otel?: OresOtelSupabasePort;
  /** Canonical RxJS lifecycle bus from `@oresoftware/ores-dnd/reactive`. */
  readonly lifecycle?: OresDndReactiveBus;
  /** Payload-free Supabase synchronization receipts. */
  readonly sync?: OresDndSupabaseSyncBus;
}

/** Hot, non-replaying RxJS stream of payload-free synchronization receipts. */
export class OresDndSupabaseSyncBus {
  readonly #receipts = new Subject<DndSupabaseSyncReceipt>();

  readonly receipts$: Observable<DndSupabaseSyncReceipt> = this.#receipts.asObservable();
  readonly failures$: Observable<DndSupabaseSyncReceipt> = this.receipts$.pipe(
    filter((receipt) => !receipt.ok),
  );

  publish(receipt: DndSupabaseSyncReceipt): void {
    this.#receipts.next(receipt);
  }

  complete(): void {
    this.#receipts.complete();
  }
}

function receipt(
  result: DndDropResult,
  channel: DndSyncChannel,
  ok: boolean,
): DndSupabaseSyncReceipt {
  const base = {
    dragId: result.dragId,
    channel,
    backend: "supabase" as const,
    ok,
  };
  const targeted = result.targetId === undefined ? base : { ...base, targetId: result.targetId };
  return ok ? targeted : { ...targeted, errorCode: "sync-failed" as const };
}

async function syncOpto(
  port: OptoSyncSupabasePort,
  envelope: DndEnvelope,
  result: DndDropResult,
  bus?: OresDndSupabaseSyncBus,
): Promise<void> {
  try {
    await port.syncAcceptedDropToSupabase(envelope, result);
    bus?.publish(receipt(result, "opto-sync", true));
  } catch (error) {
    bus?.publish(receipt(result, "opto-sync", false));
    throw error;
  }
}

async function syncOtel(
  port: OresOtelSupabasePort,
  event: DndTelemetryEvent,
  result: DndDropResult,
  bus?: OresDndSupabaseSyncBus,
): Promise<void> {
  try {
    await port.syncDndEventToSupabase(event);
    bus?.publish(receipt(result, "ores-otel", true));
  } catch (error) {
    bus?.publish(receipt(result, "ores-otel", false));
    throw error;
  }
}

/**
 * Canonical accepted-drop effects plus explicit Opto-Sync and ORES-OTel
 * Supabase calls. ores-dnd owns no Supabase SDK, endpoint, table, or credential.
 *
 * Order is fail-closed:
 * forms -> Opto-Sync local -> Opto-Sync Supabase -> reactive drop/telemetry ->
 * ORES-OTel local -> ORES-OTel Supabase.
 */
export async function commitAcceptedDropWithSupabase(
  envelope: DndEnvelope,
  result: DndDropResult,
  ports: SupabaseDropCommitPorts,
): Promise<void> {
  const basePorts: DropCommitPorts = {};
  if (ports.forms !== undefined) basePorts.forms = ports.forms;
  if (ports.optoSync !== undefined) basePorts.optoSync = ports.optoSync;

  // Reuse canonical validation/local effects but hold OTel until Opto-Sync's
  // Supabase write has succeeded.
  await commitAcceptedDrop(envelope, result, basePorts);
  if (!result.accepted) return;

  const operation = result.operation;
  if (operation === undefined) {
    throw new Error("accepted drop requires an operation");
  }

  if (ports.optoSync !== undefined) {
    await syncOpto(ports.optoSync, envelope, result, ports.sync);
  }

  ports.lifecycle?.emit("drop", envelope, {
    operation,
    ...(result.targetId === undefined ? {} : { targetId: result.targetId }),
  });

  const telemetry = telemetryFor("drop", envelope, operation, result.targetId);
  if (ports.otel !== undefined) {
    await ports.otel.emitDndEvent(telemetry);
    await syncOtel(ports.otel, telemetry, result, ports.sync);
  }
}

File: src/rust/src/supabase_sync.rs
use std::convert::Infallible;

use rxrust::prelude::*;

use crate::{
    commit_accepted_drop, telemetry_for, DndDropResult, DndEnvelope, DndError,
    DndLifecyclePhase, DndOperation, DndTelemetryEvent, DropCommitPorts, OptoSyncPort,
    OresFormsPort, OresOtelPort,
};
use crate::reactive::{DndLocalEventSubject, DndReactiveEvent};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DndSyncChannel {
    OptoSync,
    OresOtel,
}

impl DndSyncChannel {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::OptoSync => "opto-sync",
            Self::OresOtel => "ores-otel",
        }
    }
}

/// Payload-free receipt safe for reactive UI/state composition.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DndSupabaseSyncReceipt {
    pub drag_id: String,
    pub channel: &'static str,
    pub backend: &'static str,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_code: Option<&'static str>,
}

pub trait OptoSyncSupabasePort: OptoSyncPort {
    fn sync_accepted_drop_to_supabase(
        &self,
        envelope: &DndEnvelope,
        result: &DndDropResult,
    ) -> Result<(), DndError>;
}

pub trait OresOtelSupabasePort: OresOtelPort {
    fn sync_dnd_event_to_supabase(&self, event: &DndTelemetryEvent) -> Result<(), DndError>;
}

pub type DndLocalSyncSubject =
    rxrust::subject::LocalSubject<'static, DndSupabaseSyncReceipt, Infallible>;

#[must_use]
pub fn local_sync_subject() -> DndLocalSyncSubject {
    Local::subject::<DndSupabaseSyncReceipt, Infallible>()
}

pub struct SupabaseDropCommitPorts<'a> {
    pub forms: Option<&'a dyn OresFormsPort>,
    pub opto_sync: Option<&'a dyn OptoSyncSupabasePort>,
    pub otel: Option<&'a dyn OresOtelSupabasePort>,
    pub lifecycle: Option<&'a mut DndLocalEventSubject>,
    pub sync: Option<&'a mut DndLocalSyncSubject>,
}

struct OptoBaseAdapter<'a>(&'a dyn OptoSyncSupabasePort);

impl OptoSyncPort for OptoBaseAdapter<'_> {
    fn persist_accepted_drop(
        &self,
        envelope: &DndEnvelope,
        result: &DndDropResult,
    ) -> Result<(), DndError> {
        self.0.persist_accepted_drop(envelope, result)
    }
}

fn receipt(
    result: &DndDropResult,
    channel: DndSyncChannel,
    ok: bool,
) -> DndSupabaseSyncReceipt {
    DndSupabaseSyncReceipt {
        drag_id: result.drag_id.clone(),
        channel: channel.as_str(),
        backend: "supabase",
        ok,
        target_id: result.target_id.clone(),
        error_code: (!ok).then_some("sync-failed"),
    }
}

fn publish_sync(
    sync: Option<&mut DndLocalSyncSubject>,
    value: DndSupabaseSyncReceipt,
) {
    if let Some(sync) = sync {
        sync.next(value);
    }
}

/// Runs canonical accepted-drop validation/local effects and then calls the
/// injected Opto-Sync and ORES-OTel Supabase methods.
///
/// No Supabase SDK, URL, table or credential is owned by ores-dnd.
pub fn commit_accepted_drop_with_supabase(
    envelope: &DndEnvelope,
    result: &DndDropResult,
    mut ports: SupabaseDropCommitPorts<'_>,
) -> Result<(), DndError> {
    let opto_base = ports.opto_sync.map(OptoBaseAdapter);
    commit_accepted_drop(
        envelope,
        result,
        DropCommitPorts {
            forms: ports.forms,
            opto_sync: opto_base
                .as_ref()
                .map(|adapter| adapter as &dyn OptoSyncPort),
            otel: None,
        },
    )?;
    if !result.accepted {
        return Ok(());
    }

    let operation = result
        .operation
        .ok_or_else(|| DndError("accepted drop requires an operation".to_owned()))?;

    if let Some(opto_sync) = ports.opto_sync {
        match opto_sync.sync_accepted_drop_to_supabase(envelope, result) {
            Ok(()) => publish_sync(
                ports.sync.as_deref_mut(),
                receipt(result, DndSyncChannel::OptoSync, true),
            ),
            Err(error) => {
                publish_sync(
                    ports.sync.as_deref_mut(),
                    receipt(result, DndSyncChannel::OptoSync, false),
                );
                return Err(error);
            }
        }
    }

    if let Some(lifecycle) = ports.lifecycle.as_deref_mut() {
        lifecycle.next(DndReactiveEvent::new(
            DndLifecyclePhase::Drop,
            envelope.clone(),
            Some(operation),
            result.target_id.clone(),
        )?);
    }

    let telemetry = telemetry_for(
        DndLifecyclePhase::Drop,
        envelope,
        Some(operation),
        result.target_id.clone(),
    );
    if let Some(otel) = ports.otel {
        otel.emit_dnd_event(&telemetry)?;
        match otel.sync_dnd_event_to_supabase(&telemetry) {
            Ok(()) => publish_sync(
                ports.sync.as_deref_mut(),
                receipt(result, DndSyncChannel::OresOtel, true),
            ),
            Err(error) => {
                publish_sync(
                    ports.sync.as_deref_mut(),
                    receipt(result, DndSyncChannel::OresOtel, false),
                );
                return Err(error);
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::{cell::RefCell, rc::Rc};

    use super::*;
    use crate::{decode_envelope_json, ValidationOptions};

    const VALID: &str =
        include_str!("../../../contracts/instances/DndEnvelope/valid/text-copy.json");

    struct OptoFake(Rc<RefCell<Vec<&'static str>>>);

    impl OptoSyncPort for OptoFake {
        fn persist_accepted_drop(
            &self,
            _envelope: &DndEnvelope,
            _result: &DndDropResult,
        ) -> Result<(), DndError> {
            self.0.borrow_mut().push("opto-local");
            Ok(())
        }
    }

    impl OptoSyncSupabasePort for OptoFake {
        fn sync_accepted_drop_to_supabase(
            &self,
            _envelope: &DndEnvelope,
            _result: &DndDropResult,
        ) -> Result<(), DndError> {
            self.0.borrow_mut().push("opto-supabase");
            Ok(())
        }
    }

    struct OtelFake(Rc<RefCell<Vec<&'static str>>>);

    impl OresOtelPort for OtelFake {
        fn emit_dnd_event(&self, event: &DndTelemetryEvent) -> Result<(), DndError> {
            assert!(!serde_json::to_string(event)?.contains("hello"));
            self.0.borrow_mut().push("otel-local");
            Ok(())
        }
    }

    impl OresOtelSupabasePort for OtelFake {
        fn sync_dnd_event_to_supabase(
            &self,
            event: &DndTelemetryEvent,
        ) -> Result<(), DndError> {
            assert!(!serde_json::to_string(event)?.contains("hello"));
            self.0.borrow_mut().push("otel-supabase");
            Ok(())
        }
    }

    #[test]
    fn both_supabase_hooks_run_after_local_opto_and_keep_payload_out_of_receipts() {
        let decoded = decode_envelope_json(VALID, ValidationOptions::default());
        assert!(decoded.is_ok());
        let Ok(envelope) = decoded else {
            return;
        };
        let calls = Rc::new(RefCell::new(Vec::new()));
        let opto = OptoFake(Rc::clone(&calls));
        let otel = OtelFake(Rc::clone(&calls));
        let mut lifecycle = crate::reactive::local_event_subject();
        let mut sync = local_sync_subject();
        let receipts = Rc::new(RefCell::new(Vec::new()));
        let receipt_sink = Rc::clone(&receipts);
        sync.clone().subscribe(move |value| receipt_sink.borrow_mut().push(value));

        let result = DndDropResult {
            drag_id: envelope.drag_id.clone(),
            accepted: true,
            operation: Some(DndOperation::Copy),
            target_id: Some("field-1".to_owned()),
            error_code: None,
        };
        let committed = commit_accepted_drop_with_supabase(
            &envelope,
            &result,
            SupabaseDropCommitPorts {
                forms: None,
                opto_sync: Some(&opto),
                otel: Some(&otel),
                lifecycle: Some(&mut lifecycle),
                sync: Some(&mut sync),
            },
        );
        assert!(committed.is_ok());
        assert_eq!(
            calls.borrow().as_slice(),
            ["opto-local", "opto-supabase", "otel-local", "otel-supabase"]
        );
        let json = serde_json::to_string(&*receipts.borrow());
        assert!(json.is_ok());
        assert!(!json.unwrap_or_default().contains("hello"));
        assert_eq!(receipts.borrow().len(), 2);
    }
}
